### PR TITLE
DM-5660: Add motivated model fits to validate_drp photometric and astrometric scatter/repeatability analysis and plots

### PR DESCRIPTION
**Pull request by wmwv**
Implements well-motivated theoretical fits to the astrometric and photometric performance measurements based on derivations from LSST Overview paper.
http://arxiv.org/pdf/0805.2366v4.pdf
Photometric errors described by
Eq. 5
sigma_(rand) = (0.039 - gamma) \* x + gamma \* x^2 [mag^2]
where x = 10^(0.4*(m-m_5))
Eq. 4
sigma_1^2 = sigma_sys^2 + sigma_rand^2
Astrometric Errors 
error = C \* theta / SNR
Based on helpful comments from Zeljko Ivezic
"""
I think eq. 5 from the overview paper (with gamma = 0.039 and m5 = 24.35;
the former I assumed and the latter I got from the value of your 
analytic fit that gives err=0.2 mag) would be a much better fit than
the adopted function for mag < 21 (and it is derived from first principles). 
Actually, if you fit for the systematic term (eq. 4) and gamma and m5,
it would be a nice check whether there is any “weird” behavior in 
analyzed data (and you get the limiting depth, m5, even if you don’t
go all the way to the faint end).
Similarly, for the astrometric random errors, we’d expect 
error = C \* theta / SNR,
where theta is the seeing (or a fit parameter), SNR is the photometric
SNR (i.e. 1/err in mag), and C ~ 1 (empirically, and 0.6 for the idealized
maximum likelihood solution and gaussian seeing). 
"""

### CODE REVIEW

**Comment by danielsf** [comment]
Just sticking my nose where it doesn't belong:

If it would help,the calcSNR_m5 method here

https://github.com/lsst/sims_photUtils/blob/master/python/lsst/sims/photUtils/SignalToNoise.py

can calculate the photometric signal to noise given m5 and a few system parameters and throughputs.  It has been checked to be consistent with equations 4 and 5 in the overview paper.  There is also a method calcGamma to calculate the appropriate value of gamma given m5.

Take it or leave it.  I just wanted to make sure you were aware that these methods existed.

**Comment by wmwv** [comment]
@danielsf   Thanks, I didn't know about that and very much appreciate your note!

I unfortunately don't think I can use that here as `sims_photUtils` has some substantial dependencies that I don't want to add here:

```
setupRequired(sims_sed_library)
setupRequired(sims_maps)
setupRequired(sims_utils)
setupRequired(throughputs)
```

If it would be possible to expose some of these utility functions without requiring the large and substantial `sims_sed_library` and `sims_maps` then I might be interested in adopting some of these routines this to ensure consistency.

**Comment by danielsf** [comment]
@wmv That (stripping out the massive data file dependencies from sims_photUtils) is on our list of things to do in the near future (I guess we hadn't considered removing sims_maps; definitely sims_sed_library).  I will ping you when it is accomplished.

File: README.md [hunk 1]
```diff
@@ -1,20 +1,21 @@
 Validate an LSST DM processCcd.py output repository
 against a set of LSST Science Requirements Document Key Performance Metrics.
+and expected performance following the LSST Overview paper.
```
Comment (jonathansick) [inline README.md]: Typo: move this into previous sentence.

File: README.md [hunk 2]
```diff
@@ -1,20 +1,21 @@
...
 validateDrp.py CFHT/output
 ```
 
-will produces output, plots, JSON files analyzing the processed data in `CFHT/output`.  Metrics will be separately calculated for each filter represented in the repository.  Replace `CFHT/output` with your favorite processed data repository and you will get reasonable output.
+will produces output, plots, JSON files analyzing the processed data in `CFHT/output`.  Metrics will be separately calculated for each filter represented in the repository.  
+
+Replace `CFHT/output` with your favorite processed data repository and you will get reasonable output.
 
 One can run `validateDrp.py` in any of the following modes:
 1. use no configuration file (as above)
```
Comment (jonathansick) [inline README.md:16]: If you want this to render as a list in Markdown you can add a blank line before the first list item.

File: python/lsst/validate/drp/check.py [hunk 3]
```diff
@@ -21,8 +21,10 @@
 from __future__ import print_function, division
 
 import numpy as np
+from scipy.optimize import curve_fit
```
Comment (jonathansick) [inline python/lsst/validate/drp/check.py:24]: I'm not a Stack developer, but do you need to add `scipy` to `ups/validate_drp.table` now? Same for `matplotlib` actually.

File: python/lsst/validate/drp/check.py [hunk 4]
```diff
@@ -65,6 +67,44 @@ def magNormDiff(cat):
     return normDiff
 
 
+def fitExp(x, y, y_err, deg=2):
+    """Fit an exponential quadratic to x, y, y_err.
+    """
+    fit_params, fit_param_covariance = \
+        curve_fit(expModel, x, y, p0=[1, 0.02, 5], sigma=y_err)
```
Comment (jonathansick) [inline python/lsst/validate/drp/check.py:82]: I can't find where `expModel` is defined.

File: python/lsst/validate/drp/check.py [hunk 5]
```diff
@@ -151,9 +199,11 @@ def checkPhotometry(snr, mag, mmagrms, dist, match,
     snr : list or numpy.array
         Median SNR of PSF flux
     mag : list or numpy.array
-        Average magnitudes of each star
+        Average magnitudes of each star.  [mag]
+    mmagErr : list or numpy.array
+        Uncertainties in magnitudes of each star.  [mmag]
     mmagrms ; list or numpy.array
```
Comment (jonathansick) [inline python/lsst/validate/drp/check.py]: Should be a colon here.

File: python/lsst/validate/drp/check.py (hunk 3 above)
Comment (timj) [inline python/lsst/validate/drp/check.py:24]: Yes and yes.

File: python/lsst/validate/drp/check.py [hunk 6]
```diff
@@ -181,15 +233,80 @@ def checkPhotometry(snr, mag, mmagrms, dist, match,
...
+    return pipeBase.Struct(name="checkPhotometry",
+                           photRmsScatter=photScatter, 
+                           photErrModel=photErrModel.__doc__,
+                           photErrFitParams=fit_params)
+
+
+def astromErrModel(snr, theta=1, sigma_sys=0.01, C=1):
+    """Calculate expected astrometric uncertainty based on SNR.
+
+    mas = C*theta/SNR
```
Comment (jonathansick) [inline python/lsst/validate/drp/check.py]: Interesting call whether math like this should be written ascii-style for clarity to a code reader to rendered as LaTeX (below) for a future doc reader. We don't have a policy on this, so no need to change this yet.

```
.. math:: \sigma_\mathrm{astrom} = \frac{C \theta}{\mathcal{S/N}}
```

File: python/lsst/validate/drp/plot.py [hunk 7]
```diff
@@ -157,6 +170,95 @@ def plotExpFit(x, y, y_err, deg=2, ax=None, verbose=False):
...
+    Parameters
+    ----------
+    snr : list or numpy.array
+        S/N of photometric measurements
+    dist : list or numpy.array
+        Separation from reference [mas]
+
+    Returns
+    -------
+    float
```
Comment (jonathansick) [inline python/lsst/validate/drp/plot.py]: I don't think you mean for the plot function to return anything

File: python/lsst/validate/drp/check.py [hunk 8]
```diff
@@ -181,15 +233,80 @@ def checkPhotometry(snr, mag, mmagrms, dist, match,
...
+
+    Parameters
+    ----------
+    snr : list or numpy.array
+        S/N of photometric measurements
+    theta : float or numpy.array, optional
+        Seeing [mas]
```
Comment (jonathansick) [inline python/lsst/validate/drp/check.py]: Are the units of seeing really milliarcsec if the default/fitting seed is 1?

File: python/lsst/validate/drp/check.py [hunk 9]
```diff
@@ -181,15 +233,80 @@ def checkPhotometry(snr, mag, mmagrms, dist, match,
...
+    sigma_sys : float
+        Systematic error floor [mas]
+    C : float
+        Scaling factor
+    verbose : bool, optional
```
Comment (jonathansick) [inline python/lsst/validate/drp/check.py]: Doesn't belong here.

File: python/lsst/validate/drp/check.py [hunk 10]
```diff
@@ -181,15 +233,80 @@ def checkPhotometry(snr, mag, mmagrms, dist, match,
...
+def photErrModel(mag, sigmaSys, gamma, m5):
+    """Fit model of photometric error from LSST Overview paper
+    http://arxiv.org/abs/0805.2366v4
+
+    Photometric errors described by
+    Eq. 4
+    sigma_1^2 = sigma_sys^2 + sigma_rand^2
+
+    Eq. 5
+    sigma_(rand) = (0.04 - gamma) * x + gamma * x^2 [mag^2]
```
Comment (jonathansick) [inline python/lsst/validate/drp/check.py]: sigma_(rand)^2

File: python/lsst/validate/drp/plot.py [hunk 11]
```diff
@@ -157,6 +170,95 @@ def plotExpFit(x, y, y_err, deg=2, ax=None, verbose=False):
...
+
+    ax.plot(x_model, fit_model_mas_err,
+            color=color, linewidth=2,
+            label=label)
+    # Set the x limits back to their original values.
+    ax.set_xlim(xlim)
+
+
+def plotPhotErrModelFit(mag, mmag_err, fit_params=None, color='red', ax=None, verbose=True):
+    """plot model of photometric error from LSST Overview paper
```
Comment (jonathansick) [inline python/lsst/validate/drp/plot.py]: Add a ref "(Eq 4 and 5)" here to help someone in the future.

File: python/lsst/validate/drp/check.py (hunk 6 above)
Comment (jonathansick) [inline python/lsst/validate/drp/check.py]: Also add sigma_sys to this equation.

**Comment by jonathansick** [comment]
Aside from the issues mentioned in the line notes this all seems okay. I haven't independently run the code to check the plots, though.

I don't have a good feel for how robust these fits are. If there's a weird dataset is it possible for the fits to go crazy? I'm not sure it's necessary to add in either code to check the robustness of the fits (even if it's a basic bootstrap); depends on the use cases of the fitted values.

**Comment by jonathansick** [comment]
Also, it might be useful to capture the provenance of the model for astrometric scatter somewhere in the docstrings.

File: python/lsst/validate/drp/check.py (hunk 3 above)
Comment (wmwv) [inline python/lsst/validate/drp/check.py:24]: Ah, thank you!  Good catch.

File: python/lsst/validate/drp/check.py (hunk 4 above)
Comment (wmwv) [inline python/lsst/validate/drp/check.py:82]: Thanks for asking this.  I've cleaned things up correctly now.

expModel was defined plot.py.  It should have been (and is now) in check.py

But fitExp was only used in plot.py, so this worked by the time fitExp was called.

**Pull request merged**